File: backend/app/Models/Transaction.php
<?php

namespace App\Models;

use App\Traits\HasTenant;
use Database\Factories\TransactionFactory;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

class Transaction extends Model
{
    use HasFactory, HasTenant;

    /**
     * The attributes that are mass assignable.
     *
     * @var array<int, string>
     */
    protected $fillable = [
        'organization_id',
        'account_id',
        'category_id',
        'type',
        'amount',
        'occurred_at',
        'description',
        'attachment_path',
    ];

    /**
     * Get the attributes that should be cast.
     *
     * @return array<string, string>
     */
    protected function casts(): array
    {
        return [
            'amount' => 'decimal:2',
            'occurred_at' => 'datetime',
        ];
    }

    /**
     * The organization that owns the transaction.
     */
    public function organization(): BelongsTo
    {
        return $this->belongsTo(Organization::class);
    }

    /**
     * The account that owns the transaction.
     */
    public function account(): BelongsTo
    {
        return $this->belongsTo(Account::class);
    }

    /**
     * The category that owns the transaction.
     */
    public function category(): BelongsTo
    {
        return $this->belongsTo(Category::class);
    }

    /**
     * Create a new factory instance for the model.
     */
    protected static function newFactory()
    {
        return TransactionFactory::new();
    }
}

File: app/lib/features/transactions/data/models/transaction.dart
class Transaction {
  final int id;
  final String description;
  final double amount;
  final String type; // 'income' or 'expense'
  final DateTime occurredAt;
  final int? accountId;
  final String? accountName;
  final int? categoryId;
  final String? categoryName;
  final DateTime createdAt;
  final DateTime updatedAt;

  Transaction({
    required this.id,
    required this.description,
    required this.amount,
    required this.type,
    required this.occurredAt,
    this.accountId,
    this.accountName,
    this.categoryId,
    this.categoryName,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Transaction.fromJson(Map<String, dynamic> json) {
    return Transaction(
      id: json['id'] as int,
      description: json['description'] as String,
      amount: (json['amount'] as num).toDouble(),
      type: json['type'] as String,
      occurredAt: DateTime.parse(json['occurred_at'] as String),
      accountId: json['account']?['id'] as int?,
      accountName: json['account']?['name'] as String?,
      categoryId: json['category']?['id'] as int?,
      categoryName: json['category']?['name'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }
}

